File: deps/PicklingTools170Release/Java/com/picklingtools/midas/MidasTalker_ex2.java

// Longer example showing how to handle misconnects, clients going away,
// etc.

package com.picklingtools.midas;


import java.io.IOException;
import java.net.ConnectException;
import java.net.UnknownHostException;
import java.text.ParseException;

import com.picklingtools.pythonesque.Tab;
import com.picklingtools.midas.MidasSocket_.*;  // get all the nested enums

public class MidasTalker_ex2 {

    // Main loop where you send and receive messages to the MidasTalker mt
    public static void sendloop (MidasTalker mt) throws IOException {
	while (true) {

	    // Create a message (a table) and send it to the server
	    // (Server may be OpalPythonDaemon, OpalDaemon, or 
	    // MidasServer (C++ or Python)
	    try {
		Tab shared = new Tab("{ 'shared': 1 }");
		Tab t=new Tab("{'a':1,'hello':[3,3,4.4,5.5],'nest':{'cc':17}}");
		t.put("proxy1", shared);
		t.put("proxy2", shared);  // shared data structure

		t.prettyPrint(System.out);
		mt.send(t);

		// See if we can receive, wait 5 seconds
		while (true) {
		    Object res = mt.recv(5.0);
		    if (res == null) {
			System.out.println("...retrying to receive after 5 seconds...");
			// Maybe try some other work
			continue;
		    } else {
			// "Do something" with result
			Tab r = (Tab)res;
			System.out.println( "Got result:"+res);
			if (r.get("proxy1") != r.get("proxy2")) {
			    System.out.println("NOT Same"); 
			}
			break;
		    }
		}
	    } catch (ParseException pe) { 
		System.out.println("?? Illegal table?? "+pe);
	    }
	}
	    
    }

    public static void softsleep(double seconds) {
	try {
	    int milliseconds = (int)(seconds * 1000.0);
	    java.lang.Thread.sleep(milliseconds);
	} catch (InterruptedException ex) {
	    ;
	}
    }

    public static void openup(MidasTalker mt) {

	// Open the socket, with a retry in case it's not there right away
	while (true) {
	    try {
		mt.open();
		System.out.println("Opened connection to host:"+mt.host()+" port:"+mt.port());
		break;
	    } catch (UnknownHostException e) {
		System.out.println("Had trouble with name: Did you mistype the hostname?"+e);	    
		System.out.println("...couldn't open right away, backing off 5 seconds");
		softsleep(5.0);
	    } catch (ConnectException e) {
		System.out.println("Couldn't connect: Is server up?"+e);     
		System.out.println("...couldn't open right away, backing off 5 seconds");
		softsleep(5.0);

	    } catch (Exception e) { // generic problem
		System.out.println(e.getClass().getName());
		System.out.println("...couldn't open right away, backing off 5 seconds");
		softsleep(5.0);
		System.out.println(e);
	    }
	}
    }

    public static void main(String [] args) throws IOException {

	// Pass in args
	int len = args.length;
	if (len>5 || len<2 ) {
	    System.out.println("Usage: MidasTalker_ex2 hostname portnumber [ser sock arrd]");
	    System.exit(1);
	}

	String host = args[0];
	int port    = Integer.parseInt(args[1]);
	Serialization ser= len<3 ? Serialization.SERIALIZE_P2 : 
	    Serialization.get(Integer.parseInt(args[2]));
	SocketDuplex sock=len<4 ? SocketDuplex.DUAL_SOCKET : 
	    SocketDuplex.get(Integer.parseInt(args[3]));
	ArrayDisposition arrd=len<5 ? ArrayDisposition.AS_LIST : 
	    ArrayDisposition.get(Integer.parseInt(args[4]));
        	// MidasSocket_.AS_PYTHON_ARRAY is the other option

	System.out.println("Host:"+host+" port:"+port+" ser:"+ser+" sock:"+sock+" arrd:"+arrd);

	// Create, but don't actually connect
	MidasTalker mt = new MidasTalker(host, port,ser, sock, arrd);

	// Keep trying to reconnect and work
	while (true) {
	    openup(mt);
	    try {
		sendloop(mt);
	    } catch (Exception e) {
		System.out.println("Problem:"+e);
		System.out.println("Server appears to have gone away? Attempting to reconnect");
	    }
	}
	
    }

};

File: deps/PicklingTools170Release/Python/midastalker_ex2.py
#!/usr/bin/env python

# A more complicated example showing how to talk to OpalPythonDaemon
# and MidasServers, but still 
# (a) recover from errors
# (b) use timeouts on receieves
# (c) Choosing Python serialization options (Protocol 0 or 2, etc.)

from midastalker import * # Make sure you have this on your PYTHONPATH
                          # so the import picks this up (or copy it to cd)
import sys
import string
import time
import getopt

import numpy 
# Some testing for shoving Numeric shapeless arrays across the pipe
try :
  import Numeric
  a = Numeric.zeros( (), 'F')
  b = Numeric.zeros( (3), 'F')
except :
  a = []
  b = [0.0, 0.0, 0.0]
  
# No timeout exception in Python 2.2
if sys.version_info[0] == 2 and sys.version_info[1] > 2 :
  socket_timeout = socket.timeout
else :
  socket_timeout = IOError

  
# Main loop where you send and receive messages to the MidasTalker (mt)
def sendloop (mt) :

  loop = 0
  while 1 :
    loop += 1
    # if (loop==13) : sys.exit(1)
    
    # Create a message (a table) and send it on over to OpalPythonDaemon
    shared = { 'shared': 1 }
    t = { 'a':1, 'hello':[3.3, 4.4, 5.5], "nest":{'cc':17} ,'a':a, 'b':b}
    #print t['array']
    #print type(t['array'])
    t['proxy1'] = shared
    t['proxy2'] = shared

    if (loop % 10 == 0) :
      # t["VERYLARGE"] = "1"*((2**32)+1)
      #t["numpy"] = numpy.zeros(2**32+1, 'b')
      pass
      
    print 'Sending request:', t     
    mt.send(t)

    # See if we can receive, wait 5 seconds
    while 1:
      res = mt.recv(5) 
      if res is None :
        print '...retrying to receive after 5 seconds ...'
        # Maybe try to do some other work
        continue
      else :
        # "Do something" with the result
        print 'Got Result:', len(res) # ,res
        break



# Try to open up the MidasTalker, in case not immediately available 
def openup (mt) :

  # Open the file, with a retry in case it's not up right away
  while 1:
    try :
      mt.open(5.0)   # Timeout if can't open right away
      print 'Opened connection to host:',mt.host,' port:',mt.port
      break

    except socket.gaierror, val :
      print 'Had trouble with name: did you mistype the hostname?', val
      print "...couldn't open right away, backing off 5 seconds"
      time.sleep(5)
    except socket.herror, val :
      print 'Host error?  Check your DNS?', val
      print "...couldn't open right away, backing off 5 seconds"
      time.sleep(5)
    except socket_timeout, val :
      print 'Open TIMEOUT: is your server slow, suspended or in a weird state?', val
      print "...couldn't open right away, backing off 5 seconds"
      time.sleep(5)
    except error, val :
      print "Problem:", val
      print "...couldn't open right away, backing off 5 seconds"
      time.sleep(5)


# Dumb routine to make sure Signal USR1 is NOT ignored and does something
# even it is just an empty shell.  
def routine (signum, stackframe) : pass

if __name__ == '__main__' :

  # By putting this in, we can send a signal USR1 to the process to
  # call routine1 (instead of terminating). With this, we can force
  # a EINTR on a select system call with a kill -s USR1 process-id
  import signal
  signal.signal(signal.SIGUSR1, routine)
  
  try :
    opts,args=getopt.getopt(sys.argv[1:],[],["ser=","sock=","arrdisp="])
    if len(args)!=2 : raise error
  except :
    print "Usage: midastalker_ex2.py [--ser=0|1|2|-2|5] [--sock=1|2] [--arrdisp=0|1|2|4] hostname portnumber"
    sys.exit(1)

  host = args[0]
  port = string.atoi(args[1])
  serialization = SERIALIZE_P0  # Not fast, but backwards compatible
  dual_socket   = DUAL_SOCKET   # backwards compat, but use single if able
  array_disposition = ARRAYDISPOSITION_AS_LIST   # backwards compat
  
  for opt, val in opts :
    if   opt=="--ser" :     serialization     =  int(val) 
    elif opt=="--sock":
      options = { '1':0, '2':1, '777':777 }
      socket_duplex = options[val]
    elif opt=="--arrdisp" : array_disposition =  int(val)
    else : assert False, "unhandled option"

  # Create, but don't actually connect
  mt = MidasTalker(host,port,serialization,socket_duplex,array_disposition) 

  # Keep trying to reconnect and work
  while 1:
      openup(mt)
      try :
        sendloop(mt)
      except error, val :
        print 'Problem:',val
        print 'Server appears to have gone away?  Attempting to reconnect'



